### PR TITLE
fix: do not process actions that have already been included in a replay

### PR DESCRIPTION
**Pull request by effervescentia**
# Context

I work for a company that uses logux as a framework for building multiplayer experiences (https://voiceflow.com) and as we built out our solution we discovered a pretty nasty bug that occurs only when interleaving actions dispatched with `dispatch` with those using `dispatch.sync` which results in a replay that breaks the state.

With the help of some [debugging tools we created](https://github.com/voiceflow/logux-devtool) we were able to determine that during the replay one of the actions was effectively being processed twice which led to a corrupted state.

Here you can see the action `node.REMOVE_MANY` appearing twice in a row, upon inspection it is the same action with the same `meta.id`.
Screen Shot 2022-07-28 at 4 18 18 AM

# Edge Case

Conveniently the edge-case is described nicely by the test I wrote to validate the fix, so I'll use it for this example. Basically the replay occurred at a time when unprocessed actions had already been added to the log which resulted in a single action being processed twice in a row which corrupts the state.

```ts
store.dispatch({ type: 'ADD', value: 'a' })
equal(store.getState().value, '0a')

// this action is added to the log but not processed immediately
store.dispatch.sync({ type: 'ADD', value: 'b' }, { reasons: ['test'] })

// adding this action forces a state where replay must occur for 'b' and 'd'
store.dispatch({ type: 'ADD', value: 'c' })

// this action is added to the log but not processed because a replay is active
store.dispatch.sync({ type: 'ADD', value: 'd' }, { reasons: ['test'] })

/**
 * action 'b' is processed which results in a replay that includes action 'd'
 * action 'd' is queued and once the replay is done it is processed
 */
await delay(1)

/**
 * because action 'd' was in the log it is part of the replay
 * however it was processed again once the replay was complete
 * resulting in it being added twice
 */
equal(store.getState().value, '0abcdd')
```

# Solution

I toyed around with a couple of different solutions:
* excluding unprocessed actions from the scope of a replay
* adding something to `meta` to indicate that processing should be skipped

In the end I decided on re-using the `wait` lookup which is what is used for queuing in the first place.
If an action gets processed by a replay then the relevant `wait` entry is deleted. This works nicely with the existing logic so that when an action is done waiting for a replay it will skip processing if it is no longer in the `wait` lookup.

### CODE REVIEW

File: create-store-creator/index.js [hunk 1]
```diff
@@ -88,6 +89,8 @@ export function createStoreCreator(client, options = {}) {
     function replaceState(state, actions, pushHistory) {
       let last = actions.length === 0 ? null : actions[actions.length - 1][1]
       let newState = actions.reduceRight((prev, [action, id]) => {
+        delete wait[id]
```
Comment (effervescentia) [inline create-store-creator/index.js:92]: remove actions from the `wait` lookup as they are processed during a replay

File: create-store-creator/index.js [hunk 2]
```diff
@@ -184,17 +187,17 @@ export function createStoreCreator(client, options = {}) {
...
-    let wait = {}
-
     async function process(action, meta) {
       if (replaying) {
         wait[meta.id] = true
         await replaying
         if (wait[meta.id]) {
           delete wait[meta.id]
           await process(action, meta)
+          return true
```
Comment (effervescentia) [inline create-store-creator/index.js:197]: return a boolean indicating whether a change has occurred

File: create-store-creator/index.js [hunk 3]
```diff
@@ -244,8 +249,10 @@ export function createStoreCreator(client, options = {}) {
 
       if (!meta.dispatch) {
         let prevState = store.getState()
-        process(action, meta).then(() => {
-          emitter.emit('change', store.getState(), prevState, action, meta)
+        process(action, meta).then(didChange => {
+          if (didChange) {
```
Comment (effervescentia) [inline create-store-creator/index.js:253]: only emit the `change` event if the action was actually processed

**Comment by ai** [comment]
Thanks for amazing PR with many details.

Please excuse for a little delay, I needed to fix CI first.

**Comment by ai** [comment]
The fix was released in 0.8.3.
